test/unit/units/ssh_access_test.py

```diff
@@ -18,23 +18,70 @@ def test_migration_continues_on_error(
+                '/path/authorized_ec2_keys'
+            ],
+            [
+                '/system-root/etc/ssh/ssh_host_ecdsa_key.pub',
+                '/system-root/etc/ssh/ssh_host_ecdsa_key'
+            ]
         ]
-        main()
-        output_file = mock_get_migration_ssh_file()
-        expected_content = 'keys for azure\nkeys for aws ec2\n'
-        with open(output_file) as ssh_output_file:
-            assert expected_content == ssh_output_file.read()
-        os.remove(output_file)
+        mock_re_findall.return_value = 'HostKey /etc/ssh/ssh_host_ecdsa_key'
+        mock_path_exists.return_value = True
+        with patch('builtins.open', create=True) as mock_open:
+            mock_open_authorized_azure_keys = MagicMock(spec=io.IOBase)
+            mock_open_authorized_ec2_keys = MagicMock(spec=io.IOBase)
+            mock_open_authorized_keys = MagicMock(spec=io.IOBase)
+            mock_open_private_sshd = MagicMock(spec=io.IOBase)
+            mock_open_private_sshd = MagicMock(spec=io.IOBase)
+            mock_open_sshd = MagicMock(spec=io.IOBase)
+
+            def open_file(filename, mode='r'):
+                if filename == '/system-root/etc/ssh/sshd_config':
+                    return mock_open_private_sshd.return_value
+                if filename == '/etc/ssh/sshd_config':
+                    return mock_open_sshd.return_value
+                if filename == '/home/migration/.ssh/authorized_keys':
+                    return mock_open_authorized_keys.return_value
+                if filename == '/path/authorized_azure_keys':
+                    return mock_open_authorized_azure_keys.return_value
+                if filename == '/path/authorized_ec2_keys':
+                    return mock_open_authorized_ec2_keys.return_value
+
+            file_handle_azure_keys = mock_open_authorized_azure_keys.return_value.__enter__.return_value
+            file_handle_ec2_keys = mock_open_authorized_ec2_keys.return_value.__enter__.return_value
+            file_handle_authorized_keys = mock_open_authorized_keys.return_value.__enter__.return_value
+            file_handle_azure_keys.read.return_value = 'keys for azure\n'
+            file_handle_ec2_keys.read.return_value = 'keys for aws ec2'
+            file_handle_sshd_config = mock_open_sshd.return_value.__enter__.return_value
+            mock_open.side_effect = open_file
+            main()
+            assert mock_open.called
+            assert file_handle_authorized_keys.write.call_args_list == [
+                call('keys for azure\nkeys for aws ec2')
+            ]
+            assert mock_shutil_copy.call_args_list == [
+                call('/system-root/etc/ssh/ssh_host_ecdsa_key.pub', '/etc/ssh/'),
+                call('/system-root/etc/ssh/ssh_host_ecdsa_key', '/etc/ssh/')
+            ]
+            assert file_handle_sshd_config.write.call_args_list == [
+                call('HostKey /etc/ssh/ssh_host_ecdsa_key')
+            ]
```

Comment:
The testing of the sshd config update seems incomplete to me. Actually what we do is:

1. add a newline to be safe
2. add HostKey entries

The unit test should outline this logic:

1. Did we really start with a newline
2. Did we correctly add HostKey entries, at least two entries should be tested

If I would write the test I would go with real testing files taken from test/data. The only thing that needs to be mocked would be the Defaults method to point to the test/data instead of the system. That way the test becomes almost real and reading from test/data is a safe operation for unit testing

Reply:
Ok, I opted for no using any file to avoid unnecessary testing files. I will update the test with a file.